v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/dialectadapter/postgresql/PostgreSQLDialectAdapter.java

```diff
@@ -0,0 +1,518 @@
+                + "  ix.indisunique AS is_unique,"
+                + "  ix.indisprimary AS is_primary,"
+                + "  ix.indnkeyatts AS cardinality,"
+                + "  a.attnum AS ordinal_position,"
+                + "  t.typname AS type_name,"
+                + "  information_schema._pg_char_max_length(a.atttypid, a.atttypmod) AS type_length,"
+                + "  t.typcategory AS type_category,"
+                + "  ico.collation_name AS collation,"
+                + "  ico.pad_attribute AS pad"
+                + " FROM pg_catalog.pg_indexes ixs"
+                + "  JOIN pg_catalog.pg_class c ON c.relname = ixs.indexname"
+                + "  JOIN pg_catalog.pg_index ix ON c.oid = ix.indexrelid"
+                + "  JOIN pg_catalog.pg_attribute a ON c.oid = a.attrelid"
+                + "  JOIN pg_catalog.pg_type t ON t.oid = a.atttypid"
+                + "  LEFT OUTER JOIN pg_catalog.pg_collation co ON co.oid = ix.indcollation[a.attnum - 1]"
+                + "  LEFT OUTER JOIN information_schema.collations ico ON ico.collation_name = co.collname"
+                + " WHERE ixs.tablename = ?"
+                + "  AND ixs.schemaname NOT LIKE 'pg_%%'"
+                + "  AND ixs.schemaname NOT IN (%s)"
+                + " ORDER BY ix.indexrelid, ordinal_position ASC;",
+            EXCLUDED_SCHEMAS_STR);
+    ImmutableMap.Builder<String, ImmutableList<SourceColumnIndexInfo>> tableIndexesBuilder =
+        ImmutableMap.builder();
+    try (PreparedStatement statement = dataSource.getConnection().prepareStatement(query)) {
+      for (String table : tables) {
+        statement.setString(1, table);
+        ImmutableList.Builder<SourceColumnIndexInfo> indexInfosBuilder = ImmutableList.builder();
+        try (ResultSet resultSet = statement.executeQuery()) {
+          while (resultSet.next()) {
+            SourceColumnIndexInfo.Builder indexBuilder =
+                SourceColumnIndexInfo.builder()
+                    .setColumnName(resultSet.getString("column_name"))
+                    .setIndexName(resultSet.getString("index_name"))
+                    .setIsUnique(resultSet.getBoolean("is_unique"))
+                    .setIsPrimary(resultSet.getBoolean("is_primary"))
+                    .setCardinality(resultSet.getLong("cardinality"))
+                    .setOrdinalPosition(resultSet.getLong("ordinal_position"))
+                    .setIndexType(indexTypeFrom(resultSet.getString("type_category")));
+
+            // TODO(thiagotnunes): Collation support and string max length
+            String collation = resultSet.getString("collation");
+            if (collation != null) {
+              Integer typeLength = resultSet.getInt("type_length");
+              if (resultSet.wasNull()) {
+                typeLength = null;
+              }
+              boolean shouldPadSpace =
+                  isBlankPaddedType(resultSet.getString("type_name"), typeLength)
```

Comment:
I think we care only about pad-space. And my perspective could be clouded from MYSQL, so please let me know your thoughts from PG side.

In case of MYSQL blanks are ignored at all positions, so our collation detection logic creates detects blanks separately and ignores them all together.

Space (and it's variants in unicode) though is a different case. If comparison in PAD-SPACE a space at the end is ignored from comparisons, but is very much relevent at positions other than end. For example `"a\tb" < "a b"` but `"ab\t" > "ab "`
`"a\zb" = "ab" and "ab\z" = "ab"` here "\z" stands for `ctrl-z`, which in MySql world is a blank character.

What happens in PG when a blank character is compared in the middle of the string?

Reply:
A few cases in PostgreSQL, below. I think the logic is almost right, but `bpchar`, which perhaps should always be marked as `pad-space`?

```sql
-----------------------
-- Not blank padded --
-----------------------
SELECT 'a\tb' < 'a b', 'ab\t' < 'ab ';
SELECT 'a\tb'::character varying(100) < 'a b'::character varying(100), 'ab\t'::character varying(100) < 'ab '::character varying(100);
SELECT 'a\tb'::varchar(100) < 'a b'::varchar(100), 'ab\t'::varchar(100) < 'ab '::varchar(100);
SELECT 'a\tb'::text < 'a b'::text, 'ab\t'::text < 'ab '::text;
-- TRUE, TRUE

-------------------
-- Blank padded --
-------------------
SELECT 'a\tb'::character(100) < 'a b'::character(100), 'ab\t'::character(100) < 'ab '::character(100);
SELECT 'a\tb'::char(100) < 'a b'::char(100), 'ab\t'::char(100) < 'ab '::char(100);
SELECT 'a\tb'::bpchar(100) < 'a b'::bpchar(100), 'ab\t'::bpchar(100) < 'ab '::bpchar(100);
-- TRUE, FALSE (left is > than right)

---------------------------
-- Space ONLY trimmed --
---------------------------
SELECT 'ab\t'::bpchar = 'ab '::bpchar, 'ab         '::bpchar = 'ab '::bpchar;
-- FALSE, TRUE (left is equal to right)
```

Reply:
Tab is not a version of space in MySQL too (characters like non-breaking space, or [quad-space](https://www.compart.com/en/unicode/U+2001) are variants of space, but that will be more relevant when you implement the actual detection query)


I thing even char ignores trailing spaces [like](https://www.db-fiddle.com/f/idsYJsV76FWX3JQa7fKysP/0)

I am not sure if PG has any character that's a MySql style blank as in a character that's insignificant in comparisons at all positions (like \z, \0 etc)

Reply:
Sorry, I should had mentioned that the `\t` above is just representative in the query. You need to escape it like so:

```sql
SELECT E'a\t'::character(2)
```

There are only [these](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-STRINGS-ESCAPE) especial characters in PostgreSQL.

I tested all of them in db-fiddle [here](https://www.db-fiddle.com/f/4ko4mwKS6Es5tzvrxWhewd/0), alongside two unicodes you mentioned (quad-space and CTRL-Z), but they are not padded. The only character that is padded is space. 

AFAIU we only need to pad-space (trim space) for CHAR(n), CHARACTER(n) and BPCHAR(n). All others behave as expected.